plugins/modulesync.py

```diff
@@ -0,0 +1,206 @@
+            raise dnf.exceptions.Error(msg)
+        for file_path in remove_tmp_moduleyamls_files:
+            os.remove(file_path)
+
+    def _do_downloads(self, pkgs):
+        """
+        Perform the download for a list of packages
+        """
+        pkg_dict = {}
+        for pkg in pkgs:
+            pkg_dict.setdefault(str(pkg), []).append(pkg)
+
+        to_download = []
+
+        for pkg_list in pkg_dict.values():
+            pkg_list.sort(key=lambda x: (x.repo.priority, x.repo.cost))
+            to_download.append(pkg_list[0])
+        if to_download:
+            self.base.download_packages(to_download, self.base.output.progress)
+
+    def _get_packages_from_modules(self, module_spec):
+        """Gets packages from modules matching module spec
+        1. From module artifacts
+        2. From module profiles"""
+        result_query = self.base.sack.query().filterm(empty=True)
+        module_base = dnf.module.module_base.ModuleBase(self.base)
+        module_list, nsvcap = module_base.get_modules(module_spec)
+        if self.opts.newest_only:
+            module_list = self.base._moduleContainer.getLatestModules(module_list, False)
+        for module in module_list:
+            for artifact in module.getArtifacts():
+                query = self.base.sack.query(flags=hawkey.IGNORE_EXCLUDES).filterm(nevra_strict=artifact)
+                if query:
+                    result_query = result_query.union(query)
+                else:
+                    msg = _("No match for artifact '{0}' from module '{1}'").format(
+                        artifact, module.getFullIdentifier())
+                    logger.warning(msg)
+            if nsvcap.profile:
+                profiles_set = module.getProfiles(nsvcap.profile)
+            else:
+                profiles_set = module.getProfiles()
+            if profiles_set:
+                for profile in profiles_set:
+                    for pkg_name in profile.getContent():
+                        query = self.base.sack.query(flags=hawkey.IGNORE_EXCLUDES).filterm(name=pkg_name)
+                        # Prefer to add modular providers selected by argument
+                        if result_query.intersection(query):
+                            continue
+                        # Add all packages with the same name as profile described
+                        elif query:
+                            result_query = result_query.union(query)
+                        else:
+                            msg = _("No match for package name '{0}' in profile {1} from module {2}")\
+                                .format(pkg_name, profile.getName(), module.getFullIdentifier())
+                            logger.warning(msg)
+        if not module_list:
+            msg = _("No mach for argument '{}'").format(module_spec)
+            raise dnf.exceptions.Error(msg)
+
+        return result_query
+
+    def _get_providers_of_requires(self, to_test, done=None, req_dict={}):
```

Comment:
using mutable default parameter values (here `req_dict={}`) is strongly discouraged and has significant side effects (see https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/). Unless you have a reason, please use `req_dict=None`.

Reply:
Good point.

Reply:
Fixed